schedule.py

```diff
@@ -258,89 +315,161 @@ def print_results(solver, classes, professors, sections, semesters):
                 if section.semester != semester:
                     continue
                 if solver.Value(classes[(professor.name, section.name)]) == 1:
-                    if section.course in professor.preference:
+                    if professor.prefers(section.course):
                         print(professor.name + ' will be teaching ' + section.name + ' (requested)')
                     else:
                         print(professor.name + ' will be teaching ' + section.name + ' (not requested)')
         print()
 
+    # Statistics.
+    print('Statistics')
+    print('  - Number of requests met = %i' % solver.ObjectiveValue())
+    print('  - wall time       : %f s' % solver.WallTime())
+    print()
 
-def main():
-    # switch input data order
-    input_file = 'Testing data.xlsx'
-
-    semesters, sections, professors = read_input(input_file)
-
-    model, classes = create_model(professors, sections, semesters)
 
-    solver = solve_model(model, classes, professors, sections, semesters)
-
-    print_results(solver, classes, professors, sections, semesters)
+# return infos for timeslots scheduling
+# list of scheduled classes tuple (professor.name, section.name) , listed by semesters
+def get_scheduled_classes(solver, classes, professors, sections, semesters):
+    scheduled_classes = {}
+    for semester in semesters:
+        profs_classes = []
+        for _, section in sections.items():
+            if section.semester != semester:
+                continue
+            for _, professor in professors.items():
+                if solver.Value(classes[(professor.name, section.name)]) == 1:
+                    profs_classes.append((professor.name, section.name))
+                    break
+        scheduled_classes[semester] = profs_classes
+    return scheduled_classes
 
-    # FIXME previous work in progress, currently unused (and obsolete)
-    '''
-    # schedule time for each class
-    # get input from excel and create Time objects
-    # Time have start time, end time, list of 1/0 for weekdays, and list of conflicts with all time slots
-    timeslots = pd.read_excel(input_file, sheet_name='Time')
-    times = []
-    for _, rows in timeslots.iterrows():
-        info = rows.tolist()
-        t = Time(info[5], info[6], info[0:5])
-        times.append(t)
-    for t in times:
-        conflicts = {}
-        for c in times:
-            if t.conflict(c):
-                conflicts[c] = 1
-            else:
-                conflicts[c] = 0
-        t.conflicts = conflicts
 
-    # run on semester 0 only for now
-    sem_class = scheduled_classes[0]
-    sem_prof_class = prof_class[0]
+def create_timetable_model(profs_classes, professors, times):
+    prof_teach = {}  # professor to list of scheduled classes
+    for prof_sec in profs_classes:
+        prof = prof_sec[0]
+        sec = prof_sec[1]
+        if prof in prof_teach:
+            prof_teach[prof].append(sec)
+        else:
+            prof_teach[prof] = [sec]
```

Comment:
I think this is the wrong approach. Since which prof teaches which section has already determined, all you have to do is assign sections to times (which you do in the next few lines). A better approach is to have a map of section -> prof, then use that to determine preferences. This should greatly reduce the number of nested loops you have to do.

Reply:
The prof_teach is used for the constraint that avoids time conflict for one professor. Because it needs to check, for each professor, the sections don't conflict. So I set up a dictionary that links each professor to sections.
I agree that your approach should be used to determine preferences, but I think I need the prof_teach for the constraint?

Reply:
You're right - feel free to ignore my comment here.